### ISSUE TITLE
Word2Vec java.lang.ArrayIndexOutOfBoundsException in VectorCalculationsThread 0 to 3

### ISSUE THREAD

**Issue by ozborn** [deeplearning4j/deeplearning4j#1039]
Using:
>              <nd4j.version>0.4-rc3.6</nd4j.version>
>              <dl4j.version>0.4-rc3.8</dl4j.version>"

I ran into issue https://github.com/deeplearning4j/deeplearning4j/issues/846, still seen last month by lhyan792 in glitter. Since this has been fixed in master, I decided to update to master:
>                <nd4j.version>0.4-rc3.9-SNAPSHOT</nd4j.version>
>                <dl4j.version>0.4-rc3.9-SNAPSHOT</dl4j.version>

It fixed the problem, but unfortunately I ran into ArrayIndexOutofBoundsException in 4 different VectorCaluclationThreads, the first failure is below:

```
Exception in thread "VectorCalculationsThread 2" java.lang.RuntimeException: java.lang.ArrayIndexOutOfBoundsException: 1000
	at org.deeplearning4j.models.sequencevectors.SequenceVectors$VectorCalculationsThread.run(SequenceVectors.java:798)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 1000
	at org.deeplearning4j.models.embeddings.learning.impl.elements.SkipGram.iterateSample(SkipGram.java:211)
	at org.deeplearning4j.models.embeddings.learning.impl.elements.SkipGram.skipGram(SkipGram.java:124)
	at org.deeplearning4j.models.embeddings.learning.impl.elements.SkipGram.learnSequence(SkipGram.java:99)
	at org.deeplearning4j.models.sequencevectors.SequenceVectors.trainSequence(SequenceVectors.java:155)
	at org.deeplearning4j.models.sequencevectors.SequenceVectors$VectorCalculationsThread.run(SequenceVectors.java:786)
```
A larger more complete output is in this gist:
https://gist.github.com/ozborn/d05f15f5ef36c1675154

I actually had word2vec working just fine with an earlier version of word2vec, but I am reluctant to go back to that. I am running on Ubuntu Linux using openblas. Any insight appreciated.

**Comment by raver119** [comment]
I need to see gist of your code where you initialize Word2Vec

**Comment by raver119** [comment]
Actually, your bug is most probably result of adagrad(true) and negative(>0) being used.
Word2Vec AdaGrad implementation was flawed.

Right now i'm pushing fixes for it into current master.

But once again, if you'll show me gist of your source code, i'll be really thankful :)

**Comment by ozborn** [comment]
That's good news, my source code is here:
https://gist.github.com/ozborn/8419075803fc49da95e3

However I am not using AdaGrad, but I am using negative > 0.

Open to any suggestions.

**Comment by ozborn** [comment]
It just occured to me that I may be importing older classes with the same name (?), since this code was written in mid-2015 and it is possible that the classes in the gist have been updated to new packages. Below are my dl4j  imports I forgot to include in the gist above.
```
import org.deeplearning4j.models.embeddings.WeightLookupTable;
import org.deeplearning4j.models.embeddings.inmemory.InMemoryLookupTable;
import org.deeplearning4j.models.word2vec.VocabWord;
import org.deeplearning4j.models.word2vec.Word2Vec;
import org.deeplearning4j.models.word2vec.wordstore.inmemory.InMemoryLookupCache;
import org.deeplearning4j.plot.Tsne;
import org.deeplearning4j.text.sentenceiterator.FileSentenceIterator;
import org.deeplearning4j.text.sentenceiterator.SentenceIterator;
import org.deeplearning4j.text.sentenceiterator.SentencePreProcessor;
import org.deeplearning4j.text.tokenization.tokenizer.TokenPreProcess;
import org.deeplearning4j.text.tokenization.tokenizer.preprocessor.EndingPreProcessor;
import org.deeplearning4j.text.tokenization.tokenizerfactory.DefaultTokenizerFactory;
import org.deeplearning4j.text.tokenization.tokenizerfactory.TokenizerFactory;
import org.deeplearning4j.util.SerializationUtils;
import org.springframework.core.io.ClassPathResource;
```

**Comment by raver119** [comment]
Unfortunately, the only possible reason for that exception, is useAsaGrad(true) together with negative>0 with some yet unknown branch. That exception was fixed somewhere between 3.7 and 3.8 releases, however useAdaGrad(true) produced unexpected model.

Right now in current master useAdaGrad(true) is working properly, and i'll close this issue for now.
If you're going to keep playing with current master, don't forget to pull all recent changes please.

If you'll find anything else - please fill new issue here.

Thanks for help.

**Comment by raver119** [comment]
Please, give me some additional info:
- Java version
- Which JVM you're using, Oracle or OpenJDK

**Comment by ozborn** [comment]
I'm using Oracle's JDK, u60

**Comment by raver119** [comment]
Thanks for your help, i've got this bug fixed too.

**Comment by ozborn** [comment]
Thanks! 
I tested it myself as well and it works although I'm still confused as to why the calculation in SkipGram.java line 211 can give an out of bound results. My test results still look reasonable to me though.

**Comment by athawk81** [comment]
I am getting this bug presently, with version 3.8.  I am using negativeSampling but am most certainly not using Adagrad.  Has the bug fix been put in place yet in the binaries one pulls down from MavenCentral?  I am also using Oracle's JDK, Java 8.

**Comment by raver119** [comment]
Yes, it's fixed in current master, and fix will be included in 3.9 release.

Right now the only possible workaround for 3.8 is to set negative sampling to 0 via Word2Vec.Builder.negative() method
